tests/python/contrib/test_hexagon/test_launcher.py

```diff
@@ -40,6 +42,251 @@
+
+                    if sched_type == 1:
+                        pass
+                    elif sched_type == 2:
+                        sched[C].vectorize(C.op.axis[1])
+                    else:
+                        raise Exception("Unknown schedule type")
+
+                    # This module is only created so humans can inspect its IR.
+                    module_for_ir_dump = tvm.lower(sched, [A,B,C], "foo")
+
+                    report_path = os.path.join(host_output_dir, f'{version_name}.txt')
+
+                    with open(report_path, 'w') as f:
+                        f.write("LOWERED IR MODULE:\n")
+                        f.write(str(module_for_ir_dump))
+                        f.write('\n')
+
+                        target_hexagon = tvm.target.hexagon("v68", link_params=True)
+                        func = tvm.build(
+                            sched, [A, B, C], tvm.target.Target(target_hexagon, host=target_hexagon), name="add_hvx"
+                        )
+
+                        host_dso_binary_path = os.path.join(host_output_dir, f'test_binary-{version_name}.so')
+                        target_dso_binary_filename = 'test_binary.so'
+
+                        func.save(str(host_dso_binary_path))
+                        print("SAVED BINARY TO HOST PATH: {}".format(str(host_dso_binary_path)))
+
+                        if not android_serial_number:
+                            pytest.skip(msg="Skip hardware test since ANDROID_SERIAL_NUMBER is not set.")
+
+                        rpc_info = {
+                            "rpc_tracker_host": tvm_tracker_host,
+                            "rpc_tracker_port": tvm_tracker_port,
+                            "rpc_server_port": RPC_SERVER_PORT + 0,  # See note at the beginning of the file
+                            "adb_server_socket": adb_server_socket,
+                        }
+                        launcher = HexagonLauncher(serial_number=android_serial_number, rpc_info=rpc_info)
+                        launcher.upload(str(host_dso_binary_path), str(target_dso_binary_filename))
+                        launcher.start_server()
+
+                        try:
+                            with launcher.start_session() as sess:
+                                mod = launcher.load_module(target_dso_binary_filename, sess)
+
+                                host_numpy_A_data = np.ndarray(shape, dtype=dtype)
+                                host_numpy_B_data = np.ndarray(shape, dtype=dtype)
+                                host_numpy_C_data = np.ndarray(shape, dtype=dtype)
+                                host_numpy_C_data_expected = np.ndarray(shape, dtype=dtype)
+
+                                def intended_val_A(i,j):
+                                    return i + j
+
+                                def intended_val_B(i,j):
+                                    return (i+1) * (j+1)
+
+                                for i in range(shape[0]):
```

Comment:
Iteration over large arrays tends to be pretty slow in Python.  For numpy types, the iteration can be offloaded to compiled functions that operate on the entire array.

```python
j,i = np.meshgrid(np.arange(shape[1]), np.arange(shape[0]))
host_numpy_A_data = (i + j).astype(dtype)
host_numpy_B_data = ((i+1) * (j+1)).astype(dtype)
host_numpy_B_data_expected = host_numpy_A_data + host_numpy_B_data
```

Reply:
Good to know!  For people unfamiliar with that idiom, IMO the current implementation is a bit clearer and more-obviously correct.

Would it be okay to leave this as is for now, but revisit it if the performance becomes a concern?

Reply:
I think so for `host_numpy_A_data` and `host_numpy_B_data`, since the `np.meshgrid` is a bit less common of a numpy function.  For any 1-d arrays, or for computed 2-d arrays such as the `host_numpy_B_data_expected`, I think it is better to use the array-based functions.  This would remove the need for the function definition of `intended_val_A` and `intended_val_B`, since they wouldn't need to be repeated.  The result would end up being something as follows:

```python
host_numpy_A_data = np.ndarray(shape, dtype=dtype)
host_numpy_B_data = np.ndarray(shape, dtype=dtype)

for i in range(shape[0]):
    for j in range(shape[1]):
        host_numpy_A_data[i,j] = i+j
        host_numpy_B_data[i,j] = (i+1) * (j+1)
        
host_numpy_C_data_expected = host_numpy_A_data + host_numpy_B_data
host_numpy_C_data = np.zeros_like(host_numpy_C_data_expected)
```